keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,159 @@
+
+## Motivation
+This endpoint provides a good way for short-term debugging. Both tenants and cluster administrator could use this endpoint to get event stream without any changes to the cluster. Examples include:
+- allow tenants to get audit events in their own namespaces
+- provide a subset of audit events, so it will be easier to query and analyze (e.g. https://github.com/kubernetes/kubernetes/issues/56683)
+
+
+### Goals
+- Provide a new pull based audit event stream, which returns events that occurred after the stream was opened.
+- Allow users go get a subset of audit events.
+- Access control. Allow cluster administrator to get all events in cluster, allow namespace administrators to get all events in their own namespace.
+- Bandwidth limit in server scope or per-namespace.
+
+### Non-Goals
+- Access control beyond namespace granularity(e.g. allow user A to only get audits events from users A,B,C or allow user A to only get audit events of resource foo)
+- Provide a custom per-stream policies. The audit stream should be based on the cluster-wide audit policy.
+
+## User Stories
+
+### Story 1
+As a tenant, I will be able to known all things happened to my own namespace. When problem happens, I can use the audit events for trouble shooting.
+
+### Story 2
+As a developer, I will easily be able to get a subset of audit events. This will save a lot of time for debugging. I don't need to grep audit events from a couple of audit files.
+
+## Proposal
+
+### new resource endpoints
+Two new resource endpoint '/apis/audit.k8s.io/v1beta1/events' and '/apis/audit.k8s.io/v1alpha1/namespaces/{namespace}/namespacedevents' are installed.
+Unlike other resource endpoints which usually serve resource objects from key-value storage, these endpoints read events from a audit backend and only support the WATCH verb. '/apis/audit.k8s.io/v1beta1/events' provides a stream of all audit events in the cluster. '/apis/audit.k8s.io/v1alpha1/namespaces/{namespace}/namespacedevents' provides a stream of all audit events in the namespace. A new API object NamespacedEvent is introduced for the namespaced endpoint. NamespacedEvent has the same definition with the audit Event object which we already have, but `metadata.namespace` element is required for it.
+
+The audit events returned to users are the events generated according to the audit policy file. If an event is omitted by the audit policy file, the event would not be sent to user by these endpoints. For aggregated API servers, those endpoints only supports metadata-level, because kube-apiserver only record these events at metadata level.
```

Comment:
can you outline the data flow of audit information for a set of HA api servers, and how audit information from the various servers makes it to this handler, and then to the user?

Reply:
Sorry, for a HA environment, users have to know and query all HA instances of kube-apiserver.
I think this would work like a [headless services](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services).

Reply:
That doesn't work if the HA instances are behind a loadbalancer. I think it should be OK to document this as a known issue, and move forward with the current proposal to alpha. I do think HA support should be a blocker before going to beta. @liggitt @sttts WDYT?

Reply:
As for solutions, I think either the audit events need to be put into a etcd circular buffer, or the apiserver that ends up serving the request makes proxy requests to the other apiservers. That assumes apiserver <-> apiserver connectivity though, which might not be viable.

Reply:
I think we should at least have a sketch of how this would work both with multiple instances of the same apiserver and with multiple distinct servers for different API groups to decide if it seems feasible before proceeding.

Reply:
For example, would satisfying this API require all apiservers to be forwarding their audit events to a collating sink, would the filters/verbosity of that forwarding be determined by active watchers on the audit pull API, etc

Reply:
Having a pull mechanism with a collating sink sounds odd. Actually, if one really wants that, why not use the existing webhook push mechanism to send events to the sink and then pull from there?

In other words, I think having to pull all apiservers "manually" (and knowing about the server topology) is a fair compromise with a pull mechamism in HA setups.

Reply:
> why not use the existing webhook push mechanism to send events to the sink and then pull from there?


This could be an [kubernetes aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/)

Reply:
Maybe we should try prototyping this out-of-tree using the push-based dynamic webhook?

Reply:
>   prototyping this out-of-tree

This is what Aggregated API Server do, I think.